### DISCUSSION TITLE
custom action for storage backend other than local

### DISCUSSION THREAD

**Discussion by niele84-dot** [drakkan/sftpgo#1312] [Q&A] [answered]
Hi,

I have been playing with actions to modify a file upon upload and store it in my backend. I was successfully doing it with local filesystem, but for the remote one (e.g. FTP), I can't use filepaths in my script for the input and output file. 
Is it something that actions can do or they can only work with locally stored files? 
Does anyone know if there is something else in sftpgo to be used for this purpose?
Thank you!

**Comment by niele84-dot** [comment]
I think the main problem is that I'm using an FTP backend and I'm not able to find what is the temporary path where FPT-FS mounts the ftp server. I see that it is reading a temporary path, but even after setting it into the sftpgo.json, I'm not able to see it mounted and I can't therefore access it with my script.
Perhaps it is mounted on request?
Any advice ?

**Comment by drakkan** [comment]
Hello,

SFTPGo does not support FTP as a storage backend. Do you mean SFTP? Also SFTPGo doesn't mount anything.
Please clarify with an example what you want to achieve by specifying the storage backend and all relevant details. Thank you

**Reply by niele84-dot** [reply]
Hi,
yes, I meant SFTP. 
What I would like to achieve is that every time the demo user uploads a mp4 file, SFTPGo runs a script that will convert it to HEVC and upload only the latter to the SFTP server (even both will work for me).

This is what I have done.
I set up a Folder with SFTP as storage type
![image](https://github.com/drakkan/sftpgo/assets/56230147/723c564d-7d2e-4c91-9d93-b80ebee17779)
Then added it as Virtual folder  to the a "demo" user.
![image](https://github.com/drakkan/sftpgo/assets/56230147/0bbbfae6-2951-4c1d-81db-175a8f61a6ad)


I tried to set up the Event Manager in a way that for every pre-upload, it will run the converter script
![image](https://github.com/drakkan/sftpgo/assets/56230147/6cc3f15e-27bf-4037-9e8c-0601fa6d068e)

The problem is that my script doesn't know where the pre-uploaded file is stored in the local server. I tried several placeholders (e.g. {{VirtualPath}}, {{FsPath}} ) and also various Fs events (upload,pre-upload) with no success.

**Comment by drakkan** [comment, marked as answer]
pre-upload hook is executed before the uploads starts. SFTPGo upload directly to the remote SFTP server, there is no (reliable) way to access the file while the upload is in progress.

You can upload locally and configure a rule with 2 actions:

1) the first action converts the file after the upload completes
2) the second action copy the file to the SFTP folder. The copy action in the event manager abstracts the storage so it will work if the source is a local file and the destination is on another SFTP server

**Reply by niele84-dot** [reply]
thank you for the suggestion. It is working until the 2nd step. When I try to copy a file from the local storage to the SFTP folder, I get an error
`{"level":"error","time":"2023-06-08T14:17:14.463","sender":"eventmanager","message":"unable to execute action \"copy_mp4_to_ftp\" for rule \"copy_test\", elapsed 14.65574ms, err: action \"copy_mp4_to_ftp\" failed: fs copy failed for users: niele"}`

I have set up the action by using the placeholder for the VirtualPath 
<img width="946" alt="image" src="https://github.com/drakkan/sftpgo/assets/56230147/611dace2-24c8-4281-9642-16afc1511f74">


might it be because the action is executed as third?

**Reply by niele84-dot** [reply]
I tried also to copy the file to a different local folder, but It does not work

**Reply by drakkan** [reply]
the copy source path must be the path as seen by SFTPGo users not the real filesystem path

**Reply by niele84-dot** [reply]
it works. thank you!

**Reply by drakkan** [reply]
> it works. thank you!

Thank you for confirming
